### PR TITLE
Added Node Properties to template configuration.

### PR DESCRIPTION
**Pull request by kmbulebu**
Enables setting node properties, such as tool locations, on a per docker image basis.

[Review on Reviewable](https://reviewable.io/reviews/jenkinsci/docker-plugin/373)

### CODE REVIEW

File: docker-plugin/src/main/java/com/nirima/jenkins/plugins/docker/DockerTemplate.java [hunk 1]
```diff
@@ -224,6 +244,15 @@ public DockerImagePullStrategy getPullStrategy() {
     public void setPullStrategy(DockerImagePullStrategy pullStrategy) {
         this.pullStrategy = pullStrategy;
     }
+    
+    public List<? extends NodeProperty<?>> getNodeProperties() {
```
Comment (kmbulebu) [inline docker-plugin/src/main/java/com/nirima/jenkins/plugins/docker/DockerTemplate.java]: This return type is meant to match the constructor parameter type of AbstractCloudSlave. Removing the usage of the generic wildcard type here would require a cast at the super() call in DockerSlave. 

See [AbstractCloudSlave](http://javadoc.jenkins-ci.org/hudson/slaves/AbstractCloudSlave.html#AbstractCloudSlave(java.lang.String, java.lang.String, java.lang.String, int, hudson.model.Node.Mode, java.lang.String, hudson.slaves.ComputerLauncher, hudson.slaves.RetentionStrategy, java.util.List)): 
`public AbstractCloudSlave(String name,
                  String nodeDescription,
                  String remoteFS,
                  int numExecutors,
                  Node.Mode mode,
                  String labelString,
                  ComputerLauncher launcher,
                  RetentionStrategy retentionStrategy,
                  List<? extends NodeProperty<?>> nodeProperties)
                   throws Descriptor.FormException,
                          IOException
`

**Comment by kmbulebu** [comment]
I can split up readResolve, but I think it'll hurt the readability.

**Comment by KostyaSha** [comment]
You can ignore bot messages, it was my work on enhancements, but i left this plugin development. @lanwen please disable your bot.

**Comment by lanwen** [comment]
disabled. But think `readResolve` now can be splitted for 3 methods - `config1`, `config2`, `other` - so in `readResolve` will be only `if` with version selector. Big methods with lot of complexity never will be better than a couple of small and robust methods which doing one thing.

**Comment by kmbulebu** [comment]
@lanwen I pulled out the config defaults from readResolve() into a configDefaults() to reduce the complexity a bit, but I stopped short of splitting it into many methods. Hope that's okay with you.

**Comment by lanwen** [comment]
@kmbulebu thats much better, thanks!

**Comment by ndeloof** [comment]
duplicates https://github.com/jenkinsci/docker-plugin/pull/502

**Pull request merged**